### ISSUE TITLE
Investigate using the Kubernetes API directly

### ISSUE THREAD

**Issue by TomAugspurger** [TomAugspurger/kbatch#1]
The [current](https://github.com/TomAugspurger/kbatch/tree/418f20be4e06b7d0bed0c3fcba54466392d0b81e) implementation of kbatch has a client / server model. Running this involves deploying and maintaining a django application. While the application is (currently) pretty simple, it could grow more complex over time as new feature are added (like storing / streaming logs). And there are some inherent difficulties like synchronizing the status of a Job between Kubernetes and the django application.

In https://discourse.pangeo.io/t/pangeo-batch-workflows/804/12 and https://discourse.pangeo.io/t/pangeo-batch-workflows/804/14, Yuvi raised the possibility of removing the need for an intermediate server by having users directly communicate with the Kubernetes API. Advantages include

- Not having to synchronize state between multiple places
- Streaming / accessing logs is straightforward
- Hight availabiliy
- Not having to maintain yet another service

The primary challenges (for me) with this "kubernetes native" are around authentication and authorization. It seems that if we get identity management sorted out, we should be able to do things securely by only granting users permission to do things in their own namespace.

I spent some time this week digging into how this could work. https://kubernetes.io/docs/reference/access-authn-authz/authentication/ includes some details on authentication. In particular, it seems like we could use

- [Webhook Token Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication) to forward a request to the Jupyter API to authenticate a user with a Bearer token
- [OpenID Connect Tokens](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens) to authenticate with whatever backend JupyterHub is authenticating with (assuming it supports the OpenID Connect protocol; providers like Azure AD and I think Auth0).

One thing that's not at all clear to me is how to actually add an authentication method to the API server on AKS. The AKS docs, somewhat understandably given the existence of things like [managed AAD](https://docs.microsoft.com/en-us/azure/aks/managed-aad), don't seem to mention it.

I think that once authentication is sorted out, [authorization](https://kubernetes.io/docs/reference/access-authn-authz/authorization/) should be more straightforward. I think some kind of mix of ClusterRoles and RoleBindings the user's namespace should suffice.

I'll keep reading a bit about this, but will likely push a bit more on the client / server model in the meantime, since that's closer to being functional. If we *can* securely sort out the issues around identity, then I absolutely things that's the right way to go. At that point, `kbatch` would just become a small wrapper around the Kubernetes Python API that's essentially `kubectl` specialized to this workflow.

cc @yuvipanda

**Comment by yuvipanda** [comment]
YESSSSS I think removing the server will be a *huge* simplification! I think it's fundamental to the long term sustainability of the project.

The auth setups in kubernetes generally seem to require you to control the apiserver or controller manager commandline flags, so is out of the question for most hosted deployments. Something like [OpenPolicyAgent](https://www.openpolicyagent.org/) could be used to only allow the creation of certain Jobs, but that's 

My suggestion is to explore creating a very simple authenticating, authorizing proxy that is aware of the kinds of objects we want `kbatch` to access - pods, jobs, scheduledjobs and logs. So you just point the kubernetes api client to this proxy, it does authorization and authentication however it wants, forwards authorized requests to the kubernetes API. For list operations, it can enforce a label selector, and inspect responses to ensure we aren't leaking. I think this can be a generic, small project that will also be useful outside of kbatch.

**Comment by TomAugspurger** [comment]
Ah, so I thought you were suggesting `client -> Kubernetes API -> auth / identity`, which sounded complicated to implement. But instead you're suggesting `client -> "auth proxy" -> Kubernetes API`.

So then I suppose this proxy will just accept some subset of the Kubernetes API that we want to support. Users would make POST requests with the actual Kubernetes objects. And maybe we have a little CLI / Python API to make constructing those objects a bit easier.

Do you have a sense for how files should be handled? E.g. I want to execute a script like https://github.com/TomAugspurger/kbatch/blob/main/examples/ndvi-blob-storage/ndvi-job.py. Right now I upload those to blob storage, and then add an [init container](https://github.com/TomAugspurger/kbatch/blob/418f20be4e06b7d0bed0c3fcba54466392d0b81e/kbatch-server/django_kbatch_server/backend.py#L134-L149) to pull it before the pod starts. An older implementation stuck them in a ConfigMap and mounted it, but I couldn't find an easy way to clean them up.

**Comment by yuvipanda** [comment]
> So then I suppose this proxy will just accept some subset of the Kubernetes API that we want to support. Users would make POST requests with the actual Kubernetes objects. And maybe we have a little CLI / Python API to make constructing those objects a bit easier.

Yeah, ideally you can just use k8s client libraries unmodified without any issues, and get back useful error messages. As they're fairly structured, schema-ful JSON objects, this should hopefully not be too hard?

It would be useful to see if we can model everything a job needs in terms of kubernetes objects themselves. Would [Garbage Collection](https://kubernetes.io/docs/concepts/architecture/garbage-collection/) help with cleanup? Could tie the ConfigMap to the Job, although I wonder if that'll end up being cyclic - can you create the job object without the ConfigMap being present?

Another thing I want to *strongly* consider is NFS. If you're running batch jobs from a k8s cluster, and NFS is your home dir, why not just mount that in the batch job as well? This is how SLURM type things work afaict, and I think the mental model is quite reasonable. For cases where there is no NFS, configmap or object storage seem like reasonable alternatives.

**Comment by yuvipanda** [comment]
In general my hope is to avoid being responsible for storing any state ourselves, and outsource that responsibility to pre-existing infrastructure wherever possible :)

**Comment by TomAugspurger** [comment]
> It would be useful to see if we can model everything a job needs in terms of kubernetes objects themselves. Would Garbage Collection help with cleanup?

The Job `ttl_seconds_after_finished` has been very helpful for Job cleanup: https://github.com/TomAugspurger/kbatch/blob/main/kbatch-server/django_kbatch_server/backend.py#L175. I didn't see anything similar for ConfigMaps.

I haven't really thought through how to handle persistence. A small, persistent "home" directory is very convenient, just for things like scripts / notebooks (maybe with executed output) / config. I think there's some room to experiment here.

**Comment by yuvipanda** [comment]
You could mark the ConfigMap as being [owned by](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) the job, so when the job gets deleted the configmap would go too.

**Comment by TomAugspurger** [comment]
#2 took a pass at this. Some outstanding TODOs:

- [ ] Figure out how to handle files (hopefully ConfigMaps with owners)
- [ ] Write up a security model (roughly, check that every object is in the user's namespace?), verify that we implement it (https://github.com/TomAugspurger/kbatch/pull/2/files#diff-74b41e5b7dfbf279a58909b313cbeabc9d912aba6c9fe6862b1a327e0fd4754bR39)
- [ ] lots more docs
- [ ] lots more tests
- [ ] deployment (helm chart)

**Comment by TomAugspurger** [comment]
https://github.com/TomAugspurger/kbatch/pull/3 restore file handling, so users can submit a file or directory of files to be present when the job is running.

Next up is restoring the helm chart, so that I can easily deploy this as a service running on the Planetary Computer staging hub.

---

@yuvipanda do you have thoughts on whether this would be appropriate to include in the JupyterHub GitHub organization? It feels most tightly coupled to JupyterHub, much more so than Pangeo or the Planetary Computer. Would JupyterHub be open to adopting this (I'd sign on to maintain it of course)?

**Comment by yuvipanda** [comment]
AWESOME! Excited to see all this progress!

My hope is that this is not tied to JupyterHub at all - maybe auth for the proxy is pluggable? 

How about starting a `kbatch` GitHub org and moving it there to begin with? That's what I've done for github.com/notebook-sharing-space/, and we can move it around later once there is more adoption

**Comment by TomAugspurger** [comment]
> My hope is that this is not tied to JupyterHub at all - maybe auth for the proxy is pluggable?

Exactly, auth is the only thing that relies on JupyterHub right now, and that could indeed be generalized. So I'll put it in a kbatch organization for now.

**Comment by TomAugspurger** [comment]
I think we can consider this closed. Thanks for your guidance here @yuvipanda! There's still a decent amount to do, especially around user conveniences, but I'm pretty happy with where things ended up 🎉.

**Comment by yuvipanda** [comment]
SUPER EXCITED, @TomAugspurger
